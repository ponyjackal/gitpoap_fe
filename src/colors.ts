/* -------- Standard Colors ------------*/
export const White = '#FFFFFF';

/* ----------Legacy Colors------------- */
export const PrimaryBlue = '#307AE8';
export const PrimaryBlueHover = '#255092';
export const PrimaryBlueDark = '#174BE6';

export const Blue1 = '#6682E9';

export const DarkBlue2 = '#4F6693';

export const Slate1 = '#17457C';
export const Slate2 = '#8895AE';

export const Black1 = '#3C4B69';
export const Black2 = '#445371';

export const Gray1 = '#F8FAFF';
export const Gray2 = '#D7DDEA';
export const Gray3 = '#DCDDFA';
export const Gray4 = '#ACBED4';
export const Gray5 = '#E8EDFA';
export const Gray6 = '#E3EBF6';

/* -------- DeepWork Colors ------------*/
/* Proposition to use real color names, and use those when setting a theme */
export const DarkGray = '#393A51';
export const MidnightBlue = '#10111E';

export const BackgroundPanel = '#1E1F2E';

export const TextLight = '#E2E2EE';
export const TextAccent = PrimaryBlue;
export const TextGray = '#838495';
export const TextDarkGray = DarkGray;

export const ExtraHover = '#5596F6';
export const ExtraActive = '#1761CF';
export const ExtraPressed = '#1761CF';
