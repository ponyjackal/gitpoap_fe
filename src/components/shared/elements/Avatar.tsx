import styled from 'styled-components';
import { rem } from 'polished';

export const Avatar = styled.img`
  height: ${rem(80)};
  width: ${rem(80)};
  border-radius: 50%;
`;
