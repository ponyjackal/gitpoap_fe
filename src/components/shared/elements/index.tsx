export * from './Avatar';
export * from './Badge';
export * from './Button';
export * from './Checkbox';
export * from './ClaimCircle';
export * from './CollapsibleAddress';
export * from './CopyableText';
export * from './DateInput';
export * from './Divider';
export * from './GitPOAPBadge';
export * from './Header';
export * from './HexagonPath';
export * from './IconCount';
export * from './InfoHexBase';
export * from './Input';
export * from './InputWrapper';
export * from './Label';
export * from './Loader';
export * from './MultiSelect';
export * from './NavLink';
export * from './NumberInput';
export * from './OrgName';
export * from './Pagination';
export * from './POAPBadge';
export * from './Radio';
export * from './RelativeDate';
export * from './Select';
export * from './Share';
export * from './Skeletons';
export * from './Text';
export * from './TextArea';
export * from './Title';
export * from './TwitterShareButton';
