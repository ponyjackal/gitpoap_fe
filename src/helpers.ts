import { DateTime } from 'luxon';

/* Shorten check-summed version of the input address ~ 0x + 4 chars @ start + end */
export function shortenAddress(address: string, chars = 4): string {
  return `${address.substring(0, chars + 2)}…${address.substring(42 - chars)}`;
}

export const truncateAddress = (address: string, startChars = 14, endChars = 4): string => {
  if (address === '') {
    return '';
  }

  return address.slice(0, startChars) + '…' + address.slice(-endChars);
};

export const truncateString = (str: string, maxLength: number): string => {
  return str.length > maxLength ? str.slice(0, maxLength) + '…' : str;
};

export const isValidURL = (str: string): boolean => {
  const pattern = new RegExp(
    '^(https?:\\/\\/)' + // protocol
      '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' + // domain name
      '((\\d{1,3}\\.){3}\\d{1,3}))' + // OR ip (v4) address
      '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' + // port and path
      '(\\?[;&a-z\\d%_.~+=-]*)?' + // query string
      '(\\#[-a-z\\d_]*)?$',
    'i',
  ); // fragment locator

  return !!pattern.test(str);
};

// Use this helper function to adjust dates from UTC to the local timezone.
// Including the zone adjusts via timezone offset to make the date UTC.
export const formatUTCDate = (date: string): string =>
  DateTime.fromISO(date, { zone: 'utc' }).toFormat('yyyy-MM-dd');

/**
 * Github username may only contain alphanumeric characters or hyphens.
 * Github username cannot have multiple consecutive hyphens.
 * Github username cannot begin or end with a hyphen.
 * Maximum is 39 characters.
 */
export const isValidGithubHandle = (handle: string): boolean =>
  /^(?![-])(?!.*[-]{2})(?!.*[-]$)[a-zA-Z0-9-]{1,39}$/.test(handle);

/**
 * Temporary validator that prevents the GitHub handle from starting with 0x
 */
export const isValidGithubHandleWithout0x = (handle: string): boolean =>
  /^(?![-])(?!0x)(?!.*[-]{2})(?!.*[-]$)[a-zA-Z0-9-]{1,39}$/.test(handle);

export const isValidTwitterHandle = (handle: string): boolean =>
  /^[a-zA-Z0-9_]{4,15}$/.test(handle);

export const fetchWithToken = async (url: string, token: string | null) => {
  const response = await fetch(url, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  if (!response.ok || response.status >= 400) {
    throw new Error(response.statusText);
  }

  return await response.json();
};
