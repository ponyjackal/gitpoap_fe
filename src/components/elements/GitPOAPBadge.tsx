import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { TextLight, MidnightBlue } from '../../colors';

type Props = {
  imgUrl: string;
};

type HexProps = {
  size: number;
  borderSize: number;
};

const Hexagon = styled.div`
  clip-path: polygon(5% 25%, 50% 0, 95% 25%, 95% 75%, 50% 100%, 5% 75%);
`;

const HexOuterBorder = styled(Hexagon)<HexProps>`
  position: relative;
  --s: ${(props) => rem(props.size + 4 * props.borderSize)};
  width: var(--s);
  height: calc(var(--s) * 1);
  background-color: ${TextLight};
`;
const HexInnerBorder = styled(Hexagon)<HexProps>`
  --s: ${(props) => rem(props.size + 2 * props.borderSize)};
  position: absolute;
  top: ${(props) => rem(props.borderSize)};
  left: ${(props) => rem(props.borderSize)};

  width: var(--s);
  height: calc(var(--s) * 1);
  background: ${MidnightBlue};
`;

const HexBadge = styled(Hexagon)<{ url: string } & HexProps>`
  --s: ${(props) => rem(props.size)};
  position: absolute;
  top: ${(props) => rem(props.borderSize)};
  left: ${(props) => rem(props.borderSize)};

  width: var(--s);
  height: calc(var(--s) * 1);
  display: inline-block;
  font-size: initial; /* we reset the font-size if we want to add some content */
  background: url('${(props) => props.url}') no-repeat center center;
  background-size: cover;
`;

export const GitPOAPBadge = ({ imgUrl }: Props) => {
  const size = 200;
  const borderSize = 4;
  return (
    <HexOuterBorder size={size} borderSize={borderSize}>
      <HexInnerBorder size={size} borderSize={borderSize}>
        <HexBadge url={imgUrl} size={size} borderSize={borderSize} />
      </HexInnerBorder>
    </HexOuterBorder>
  );
};

/* Another option is to use SVG ~ https://stackoverflow.com/questions/67458234/mask-image-with-svg-shape-and-add-a-border*/
