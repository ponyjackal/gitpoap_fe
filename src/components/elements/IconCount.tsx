import styled from 'styled-components';
import { rem } from 'polished';
import { ExtraHover, ExtraPressed, TextGray, TextLight } from '../../colors';

type Props = {
  icon: React.ReactNode;
  href?: string;
  count: number;
};

const Count = styled.span`
  font-family: PT Mono;
  font-style: normal;
  font-weight: bold;
  font-size: ${rem(12)};
  line-height: ${rem(17)};
  letter-spacing: ${rem(0.5)};
  text-transform: uppercase;
  color: ${TextLight};
  margin-left: ${rem(6)};
`;

const Container = styled.a`
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;

  path {
    transition: 200ms fill ease;
    fill: ${TextGray};
  }

  &:hover {
    path {
      fill: ${ExtraHover};
    }
  }
  &:active {
    path {
      fill: ${ExtraPressed};
    }
  }
`;

export const IconCount = ({ icon, href, count }: Props) => {
  return (
    <Container href={href}>
      {icon}
      {count && <Count>{count}</Count>}
    </Container>
  );
};
