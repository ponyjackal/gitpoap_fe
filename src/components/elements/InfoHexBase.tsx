import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { BackgroundPanel } from '../../colors';

type Props = {
  children: React.ReactNode;
};

const StyledInfoHex = styled.div`
  display: inline-flex;
  flex-direction: column;
  min-width: ${rem(250)};
`;

const UpperTriangle = styled.div`
  clip-path: polygon(50% 0%, 0% 100%, 100% 100%);
  width: 100%;
  height: ${rem(70)};
  background-color: ${BackgroundPanel};
`;

const LowerTriangle = styled(UpperTriangle)`
  clip-path: polygon(50% 100%, 0 0, 100% 0);
`;

const Body = styled.div`
  background-color: ${BackgroundPanel};
  min-height: ${rem(100)};
`;

export const InfoHexBase = (props: Props) => {
  return (
    <StyledInfoHex>
      <UpperTriangle />
      <Body>{props.children}</Body>
      <LowerTriangle />
    </StyledInfoHex>
  );
};
