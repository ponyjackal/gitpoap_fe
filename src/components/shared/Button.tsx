import React from 'react';
import styled from 'styled-components';
import { Button as MButton } from '@mantine/core';
import { rem } from 'polished';
import { PrimaryBlue, TextAccent } from '../../colors';

type Props = {
  children: React.ReactNode;
  className?: string;
  onClick?: React.MouseEventHandler;
  disabled?: boolean;
  leftIcon?: React.ComponentProps<typeof MButton>['leftIcon'];
};

const StyledButton = styled(MButton)`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  background-color: ${PrimaryBlue};
  font-family: 'PT Mono', monospace;
  font-weight: 700;
  color: white;
  text-transform: uppercase;
  font-size: ${rem(12)};
  line-height: ${rem(18)};
  letter-spacing: ${rem(2)};
  transition: 200ms background ease;
  border-radius: ${rem(6)};
  padding: ${rem(8)} ${rem(14)};
  height: auto;

  &:hover {
    box-shadow: none;
    background-color: ${TextAccent};
  }

  &:active {
    background-color: #20457e;
  }
`;

export const Button = ({ children, className, onClick, disabled, leftIcon }: Props) => {
  return (
    <StyledButton
      compact
      className={className}
      size="md"
      disabled={disabled}
      onClick={onClick}
      leftIcon={leftIcon}
    >
      {children}
    </StyledButton>
  );
};
