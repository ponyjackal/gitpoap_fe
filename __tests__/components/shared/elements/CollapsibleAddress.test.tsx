import 'jest-styled-components';
import { CollapsibleAddress } from '../../../../src/components/shared/elements';
import { truncateAddress } from '../../../../src/helpers';
import { renderWithTheme } from '../../../__utils__/renderWithTheme';

const zeroAddress = '0x0000000000000000000000000000000000000000';

describe('CollapsibleAddress', () => {
  it('renders an CollapsibleAddress - NOT collapsed', () => {
    const { container } = renderWithTheme(
      <CollapsibleAddress address={zeroAddress} isCollapsed={false} />,
    );
    const collapse = container.firstChild;

    expect(collapse).toBeInTheDocument();
    expect(collapse).toHaveTextContent(zeroAddress);
    expect(collapse).toMatchSnapshot();
  });

  it('renders an CollapsibleAddress - collapsed', () => {
    const { container } = renderWithTheme(<CollapsibleAddress address={zeroAddress} isCollapsed />);
    const collapse = container.firstChild;

    expect(collapse).toBeInTheDocument();
    expect(collapse).toHaveTextContent(truncateAddress(zeroAddress));
    expect(collapse).toMatchSnapshot();
  });
});
