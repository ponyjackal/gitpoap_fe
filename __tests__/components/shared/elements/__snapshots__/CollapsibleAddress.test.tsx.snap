// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleAddress renders an CollapsibleAddress - NOT collapsed 1`] = `
.c0 {
  display: inline-block;
  font-family: 'PT Mono',monospace;
  font-weight: 700;
  font-size: 0.8125rem;
  line-height: 1.0625rem;
  -webkit-letter-spacing: 0.03125rem;
  -moz-letter-spacing: 0.03125rem;
  -ms-letter-spacing: 0.03125rem;
  letter-spacing: 0.03125rem;
  text-align: center;
  color: #307AE8;
  -webkit-transition: color 150ms ease-in-out;
  transition: color 150ms ease-in-out;
  cursor: pointer;
  color: #307AE8;
}

.c1:hover:not([disabled]) {
  color: #5596F6;
}

.c1:active:not([disabled]) {
  color: #1761CF;
}

.c1[disabled] {
  color: #393A51;
  cursor: not-allowed;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  font-family: PT Mono;
  font-style: normal;
  font-weight: normal;
  font-size: 0.875rem;
  line-height: 1.125rem;
  -webkit-letter-spacing: -0.03125rem;
  -moz-letter-spacing: -0.03125rem;
  -ms-letter-spacing: -0.03125rem;
  letter-spacing: -0.03125rem;
  color: #838495;
}

.emotion-1 {
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  -webkit-tap-highlight-color: transparent;
  color: inherit;
  font-size: inherit;
  line-height: 1.55;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-1:focus {
  outline-offset: 2px;
  outline: 2px solid #339af0;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<div
  class="emotion-0 c0 c1 c2 emotion-1"
>
  0x0000000000000000000000000000000000000000
  <div
    class="emotion-2"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      style="margin-left: 0.375rem;"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        height="13"
        rx="2"
        ry="2"
        width="13"
        x="9"
        y="9"
      />
      <path
        d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
      />
    </svg>
  </div>
</div>
`;

exports[`CollapsibleAddress renders an CollapsibleAddress - collapsed 1`] = `
.c0 {
  display: inline-block;
  font-family: 'PT Mono',monospace;
  font-weight: 700;
  font-size: 0.8125rem;
  line-height: 1.0625rem;
  -webkit-letter-spacing: 0.03125rem;
  -moz-letter-spacing: 0.03125rem;
  -ms-letter-spacing: 0.03125rem;
  letter-spacing: 0.03125rem;
  text-align: center;
  color: #307AE8;
  -webkit-transition: color 150ms ease-in-out;
  transition: color 150ms ease-in-out;
  cursor: pointer;
  color: #307AE8;
}

.c1:hover:not([disabled]) {
  color: #5596F6;
}

.c1:active:not([disabled]) {
  color: #1761CF;
}

.c1[disabled] {
  color: #393A51;
  cursor: not-allowed;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  font-family: PT Mono;
  font-style: normal;
  font-weight: normal;
  font-size: 0.875rem;
  line-height: 1.125rem;
  -webkit-letter-spacing: -0.03125rem;
  -moz-letter-spacing: -0.03125rem;
  -ms-letter-spacing: -0.03125rem;
  letter-spacing: -0.03125rem;
  color: #838495;
}

.emotion-1 {
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  -webkit-tap-highlight-color: transparent;
  color: inherit;
  font-size: inherit;
  line-height: 1.55;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-1:focus {
  outline-offset: 2px;
  outline: 2px solid #339af0;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<div
  class="emotion-0 c0 c1 c2 emotion-1"
>
  0x000000000000...0000
  <div
    class="emotion-2"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      style="margin-left: 0.375rem;"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        height="13"
        rx="2"
        ry="2"
        width="13"
        x="9"
        y="9"
      />
      <path
        d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
      />
    </svg>
  </div>
</div>
`;
